Source/CarthageKit/ProjectDependencyRetriever.swift

```diff
@@ -420,9 +440,43 @@ public final class ProjectDependencyRetriever {
                             )
                         }
                 }
+            }.on(terminated: {
+                if let lockFileURL = lockFileURL {
+                    ProjectDependencyRetriever.releaseLock(lockFileURL: lockFileURL)
+                }
+            })
+    }
+
+    private static func obtainLock(fileURL: URL, timeout: Int) -> SignalProducer<URL, CarthageError> {
+        //shlock -f lockfile
+        let parentURL = fileURL.deletingLastPathComponent()
+        let fileName = fileURL.lastPathComponent
+        let lockFileURL = parentURL.appendingPathComponent(".\(fileName).lock")
+        let processId = String(ProcessInfo.processInfo.processIdentifier)
+        let taskDescription = Task("/usr/bin/shlock", arguments: ["-f", lockFileURL.path, "-p", processId])
+        let retryInterval = 1
+        let retryCount = timeout == 0 ? Int.max : timeout / retryInterval
+        return SignalProducer(value: parentURL)
+            .attempt {
+                Result(at: $0, attempt: {
+                    try FileManager.default.createDirectory(at: $0, withIntermediateDirectories: true)
+                })
+            }
+            .flatMap(.merge) { _ in
+                return taskDescription.launch()
+                    .ignoreTaskData()
+                    .retry(upTo: retryCount, interval: TimeInterval(retryInterval), on: QueueScheduler(qos: .default))
+                    .mapError { _ in return .lockError(url: fileURL, timeout: timeout) }
+                    .map { _ in
+                        return lockFileURL
+                }
         }
     }
 
+    private static func releaseLock(lockFileURL: URL) {
+        _ = try? FileManager.default.removeItem(at: lockFileURL)
```

Comment:
Would it be an error if `removeItem` fails? Maybe it would be good throw here?

Reply:
I factored the FileLock out to a separate class and implemented test cases for it. The problem with adding throws is that there is no really good way to recover from the error in case the file cannot be removed.

The most obvious reason would be that the file is already gone, in which case the error is not really an error. In all other cases (e.g. file system problems) there is no real way to recover (except maybe crashing the application). The lock will be released automatically when the process dies because of the way shlock works.

Reply:
If no recovery is possible, would it make sense to crash with a message, versus a silent failure?